In [71]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [72]:
df = pd.read_csv('wk11_moreinfo.csv')

In [73]:
df

Wk      Time            Winner  Pts             Loser  Pts.1  \
0     1   7:00 PM  Florida Atlantic   43         Charlotte     13   
1     1   5:00 PM     Florida State   47          Duquesne      7   
2     1   4:00 PM          Illinois   38           Wyoming      6   
3     1  10:00 PM            Nevada   23  New Mexico State     12   
4     1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21   
..   ..       ...               ...  ...               ...    ...   
644  11   7:30 PM         (12) Utah   45           Arizona     20   
645  11   3:30 PM        Utah State   27        New Mexico     10   
646  11   3:30 PM  Washington State   52          Stanford     14   
647  11  12:00 PM  Western Kentucky   59         Charlotte      7   
648  11  12:00 PM         Wisconsin   23          Maryland     10   

              winners            losers  home_rank  away_rank  home_team_code  \
0    Florida Atlantic         Charlotte         99         99              32   
1       Florida State          Duquesne         99         99              34   
2            Illinois           Wyoming         99         99              42   
3              Nevada  New Mexico State         99         99              76   
4    Nevada-Las Vegas       Idaho State         99         99              74   
..                ...               ...        ...        ...             ...   
644              Utah           Arizona         99         99             119   
645        Utah State        New Mexico         99         99             120   
646  Washington State          Stanford         99         99             103   
647  Western Kentucky         Charlotte         99         99              21   
648         Wisconsin          Maryland         99         99             130   

     away_team_code  home_results  win_rank  lose_rank       dates  day_code  
0                37             0        99         99  2022-08-27         5  
1                49             0        99         99  2022-08-27         5  
2               221             0        99         99  2022-08-27         5  
3               121             1        99         99  2022-08-27         5  
4                76             0        99         99  2022-08-27         5  
..              ...           ...       ...        ...         ...       ...  
644              10             0        99         99  2022-11-05         5  
645             124             0        99         99  2022-11-05         5  
646             211             1        99         99  2022-11-05         5  
647             216             1        99         99  2022-11-05         5  
648             102             0        99         99  2022-11-05         5  

[649 rows x 17 columns]

In [74]:
df = df.drop(columns=['Time', 'Winner', 'Loser', 'dates'])

In [75]:
df[df['winners'] == 'Louisiana State']

Wk  Pts  Pts.1          winners             losers  home_rank  away_rank  \
124   3   65     17  Louisiana State           Southern         99         99   
206   4   31     16  Louisiana State  Mississippi State         99         99   
273   5   38      0  Louisiana State         New Mexico         99         99   
347   6   21     17  Louisiana State             Auburn         99         99   
458   8   45     35  Louisiana State            Florida         99         99   
511   9   45     20  Louisiana State        Mississippi         99          7   
614  11   32     31  Louisiana State            Alabama         99          6   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
124              53             170             0        99         99   
206              53             114             0        99         99   
273              53             124             0        99         99   
347              10              96             1        99         99   
458              31              96             1        99         99   
511              53             113             0        99          7   
614              53               3             0        99          6   

     day_code  
124         5  
206         5  
273         5  
347         5  
458         5  
511         5  
614         5

In [76]:
df[df['winners'] == 'Arkansas']

Wk  Pts  Pts.1   winners          losers  home_rank  away_rank  \
38    2   31     24  Arkansas      Cincinnati         99         99   
99    3   44     30  Arkansas  South Carolina         99         99   
182   4   38     27  Arkansas  Missouri State         99         99   
441   8   52     35  Arkansas   Brigham Young         99         99   
548  10   41     27  Arkansas          Auburn         99         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
38                7              39             0        99         99   
99                7             163             0        99         99   
182               7             116             0        99         99   
441              16              12             1        99         99   
548              10              12             1        99         99   

     day_code  
38          5  
99          5  
182         5  
441         5  
548         5

In [77]:
df = df.drop(columns=['Pts.1', 'winners', 'losers'])

In [78]:
df.tail()

Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
644  11   45         99         99             119              10   
645  11   27         99         99             120             124   
646  11   52         99         99             103             211   
647  11   59         99         99              21             216   
648  11   23         99         99             130             102   

     home_results  win_rank  lose_rank  day_code  
644             0        99         99         5  
645             0        99         99         5  
646             1        99         99         5  
647             1        99         99         5  
648             0        99         99         5

In [79]:
df

Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
0     1   43         99         99              32              37   
1     1   47         99         99              34              49   
2     1   38         99         99              42             221   
3     1   23         99         99              76             121   
4     1   52         99         99              74              76   
..   ..  ...        ...        ...             ...             ...   
644  11   45         99         99             119              10   
645  11   27         99         99             120             124   
646  11   52         99         99             103             211   
647  11   59         99         99              21             216   
648  11   23         99         99             130             102   

     home_results  win_rank  lose_rank  day_code  
0               0        99         99         5  
1               0        99         99         5  
2               0        99         99         5  
3               1        99         99         5  
4               0        99         99         5  
..            ...       ...        ...       ...  
644             0        99         99         5  
645             0        99         99         5  
646             1        99         99         5  
647             1        99         99         5  
648             0        99         99         5  

[649 rows x 10 columns]

In [80]:
y = df['Pts']
X = df.drop(columns=['Pts'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.30, random_state=2)

In [81]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

In [93]:
regr = RandomForestRegressor(max_depth=50000, random_state=0)
regr.fit(X_train, y_train)

RandomForestRegressor(max_depth=50000, random_state=0)

In [94]:
test_pred = regr.predict(X_test)

print('RMSE: '+str(np.sqrt(mean_squared_error(y_test, test_pred))))
print('R Squared: '+str(r2_score(y_test, test_pred)))
print('MSE TEST: '+str(mean_squared_error(y_test, test_pred)))
print('MAE TEST: '+str(mean_absolute_error(y_test, test_pred)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, test_pred)))

RMSE: 13.174499502730109
R Squared: -0.12839974763327455
MSE TEST: 173.5674371474359
MAE TEST: 10.59673076923077
MAPE: 0.3393666975845438


In [95]:
# MAPE: 0.3155694978397048

In [96]:
data = {'y_test': y_test, 'y_pred': test_pred}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)
new_df

y_test  y_pred   diff
297      34   34.05   0.05
134      34   34.09   0.09
413      41   40.75   0.25
129      37   37.52   0.52
595      27   27.57   0.57
..      ...     ...    ...
60       70   41.13  28.87
161      20   49.03  29.03
232      63   31.46  31.54
223      68   35.55  32.45
435      70   30.37  39.63

[195 rows x 3 columns]

A MAPE less than 5% is considered as an indication that the forecast is acceptably accurate. A MAPE greater than 10% but less than 25% indicates low, but acceptable accuracy and MAPE greater than 25% very low accuracy, so low that the forecast is not acceptable in terms of its accuracy.

In [97]:
df.head()

Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
0   1   43         99         99              32              37   
1   1   47         99         99              34              49   
2   1   38         99         99              42             221   
3   1   23         99         99              76             121   
4   1   52         99         99              74              76   

   home_results  win_rank  lose_rank  day_code  
0             0        99         99         5  
1             0        99         99         5  
2             0        99         99         5  
3             1        99         99         5  
4             0        99         99         5

In [51]:
# la = [[12, 99, 7, 12, 53, 5]]
# lsu = pd.DataFrame(la, columns=['Wk', 'home_rank', 'away_rank',
#                                 'home_team_code', 'away_team_code', 'day_code'])
# lsu

la = [[12, 99, 7, 12, 53, 1,  5]]
lsu = pd.DataFrame(la, columns=['Wk', 'home_rank', 'away_rank',
                                'home_team_code', 'away_team_code', 'home_results', 
                                'win_rank', 'lose_rank', 'day_code'])
lsu

# alabama away team code = 3
# ole miss home team code = 67

# lsu away code = 53
# arkansas home code = 12

Wk  home_rank  away_rank  home_team_code  away_team_code  day_code
0  12         99          7              12              53         5

In [52]:
regr.predict(lsu)

array([0.68])

## Update model has lsu winning (.68)

## arkansas wins
arkansas scores 39.62
lsu scores 26.12

## lsu wins
lsu 33.73
arkansas 24.26

### Results:
#### Lsu 13
#### Arkansas 10